contracts/contribute/ContributeManager.sol

```diff
@@ -314,17 +373,114 @@ contract ContributeManager is ERC721TokenReceiver {
-        INFToken(serviceRegistry).transferFrom(address(this), msg.sender, serviceId);
+        // Terminate service
+        (, uint256 terminateRefund) = IService(serviceManager).terminate(serviceId);
+
+        // Unbond service
+        (, uint256 unbondRefund) = IService(serviceManager).unbond(serviceId);
+
+        // Calculate refund
+        uint256 refund = terminateRefund + unbondRefund;
+
+        // Transfer back OLAS tokens
+        IToken(olas).transfer(contributor, refund);
+
+        // Transfer the service back to the original owner, if requested
+        if (pullService) {
+            INFToken(serviceRegistry).transferFrom(address(this), contributor, serviceId);
+        }
 
         // Zero the service info: the service is out of the contribute records, however multisig activity is still valid
         // If the same service is staked back, the multisig activity continues being tracked
-        IContributors(contributorsProxy).setServiceInfoForId(msg.sender, 0, 0, address(0), address(0));
+        IContributors(contributorsProxy).setServiceInfoForId(contributor, 0, 0, address(0), address(0));
 
-        emit Unstaked(socialId, msg.sender, serviceId, multisig, stakingInstance);
+        emit Unstaked(socialId, contributor, serviceId, multisig, stakingInstance);
+    }
+
+    /// @dev Re-stakes evicted service Id corresponding to the msg.sender.
+    function reStakeEvicted() external {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check for existing service corresponding to the social Id
+        (uint256 socialId, uint256 serviceId, address multisig, address stakingInstance) =
+            IContributors(contributorsProxy).mapAccountServiceInfo(msg.sender);
+        if (serviceId == 0) {
+            revert ServiceNotDefined(socialId);
+        }
+
+        // Check that the service is evicted
+        if (IStaking(stakingInstance).getStakingState(serviceId) != IStaking.StakingState.Evicted) {
+            revert ServiceAlreadyStaked(socialId, serviceId, multisig);
+        }
+
+        // Unstake the service
+        IStaking(stakingInstance).unstake(serviceId);
+
+        // Approve service NFT for the staking instance
+        INFToken(serviceRegistry).approve(stakingInstance, serviceId);
+
+        // Stake the service
+        IStaking(stakingInstance).stake(serviceId);
+
+        emit Restaked(socialId, msg.sender, serviceId, multisig, stakingInstance);
+
+        _locked = 1;
+    }
+
+    /// @dev Re-stakes service Id corresponding to the msg.sender from one staking instance to another.
+    /// @notice Service is unstaked, terminated, unbonded, and current service stake is returned to the contributor.
+    ///         Thus, make sure to approve a new stake amount in order to be able to re-deploy the service and stake it.
+    function reStake(address nextStakingInstance) external {
```

Comment:
Also long-awaited request from Apps team: on-click re-stake from one staking contract to another.

Reply:
Long-awaited request from Apps team: on-click re-stake of an evicted service.